node/libs/roles/src/validator/messages/consensus.rs

```diff
@@ -169,10 +172,53 @@ pub struct LeaderCommit {
     pub justification: CommitQC,
 }
 
+/// Utility for incrementally constructing a `PrepareQC` instance .
+pub struct PrepareQCBuilder {
+    instance: PrepareQC,
+    val_set_size: usize,
+}
+
+impl PrepareQCBuilder {
+    /// Create a new builder instance for a given validator set size.
+    pub fn new(val_set_size: usize) -> Self {
+        Self {
+            instance: Default::default(),
+            val_set_size,
+        }
+    }
+
+    /// Add a validator's signed message to the `PrepareQC`.
+    pub fn add(&mut self, signed_message: &Signed<ReplicaPrepare>, validator_index: usize) {
+        // TODO: refactor to cleaner code
+        if self.instance.map.contains_key(&signed_message.msg) {
+            self.instance
```

Comment:
shouldn't message signature verification happen here, so that the builder actually knows that it is building the correct thing?

Reply:
Message signature was already verified prior to aggregating. Is there a reason why that should not suffice?

Reply:
I meant to move verification here, not do it twice.

Reply:
I took the thin builder approach, delegating responsibility for correctness to the caller, to be used after all checks are done. Otherwise, there would be error checks duplications like previously, or confusion regarding who’s checking what. 

Also IMO signature should be verified even earlier than now.

Reply:
ack, please then annotate that the function expects the signature to be already verified.

Reply:
Added.